### ISSUE TITLE
[RFC] Investigate a major refactor of the option system

### ISSUE THREAD

**Issue by tarruda** [neovim/neovim#1913]
option.c currently has about 8k loc, which is a lot of code for managing editor configuration. I haven't examined that file with a great level of detail, but it seems to me that all it does is validate option names and invoke certain functions to parse and trigger events when options are set.

Maybe we can use dicts for storing options? I suggest adding a new global dict/namespace for options(prefix o:) which will be used for option storage, and an autocommand that can be used to validate/parse options. The `:set NAME` ex command will simply be a shortcut for `let o:name = value` that triggers the `OptionSet` autocommand. With this in place we can move the validation/parse logic from option.c to vimscript(which is very good at parsing/validating text).

This would be a major undertaking due to all the C -> vimscript translation(though lots of options do nothing right now and could be removed) and because all the `p_*` read/writes will have to be translated into dict lookups. I'm not sure if it would be worth all the trouble(thus the RFC), but removing that file could go a long way into improving maintenance.

**Comment by elmart** [comment]
> The :set NAME ex command will simply be a shortcut for let o:name = value

What would then be of the `let &name = ...` notation? Would we have 2 ways of setting an option in vimscript? We can't remove the `&` notation...

**Comment by tarruda** [comment]
> What would then be of the let &name = ... notation? Would we have 2 ways of setting an option in vimscript? We can't remove the & notation...

yes, I didn't mean that we should remove the existing notations, simply make them set a dict value under the hoods.

**Comment by justinmk** [comment]
IMO this is low priority

**Comment by elmart** [comment]
Well, it could be tried and tested, but I don't see a clear win here. Option module is a well-defined one (with clear functional boundaries, I mean), that can be slowly refactored towards something better. I particularly don't prefer maintaining vimscript to maintaining c very much. And performance... I don't know. Doubtful about this.

**Comment by tarruda** [comment]
> IMO this is low priority

:+1: very low priority. I just added this for discussion.

> Well, it could be tried and tested, but I don't see a clear win here. Option module is a well-defined one (with clear functional boundaries, I mean), that can be slowly refactored towards something better. I particularly don't prefer maintaining vimscript to maintaining c very much. And performance... I don't know. Doubtful about this.

Even though vimscript is not a very beautiful, it is still a very high level and garbage collected language, especially for dealing with text which seems to be a lot of what option.c does. Its hard to predict how much code could be removed by this transition, but my guess is that it would be significant. 

Another option would be to use lua once the translator is merged.

**Comment by justinmk** [comment]
Can't we embed luajit beforr the translator is done? I haven't seen much talk about that.

**Comment by tarruda** [comment]
> Can't we embed luajit beforr the translator is done? I haven't seen much talk about that.

What would be the purpose of doing that while still using the current vimscript runtime?

**Comment by justinmk** [comment]
It would allow us to write new features in lua instead of C. And support lua plugins.

**Comment by fmoralesc** [comment]
> It would allow us to write new features in lua instead of C. And support lua plugins.

:+1:

**Comment by trusktr** [comment]
How about JavaScriptjit? :D :D

**Comment by trusktr** [comment]
Or JavaScriptaot? xD xD xD

**Comment by trusktr** [comment]
I wish I could develop/add on to my favorite editor in js. :)

**Comment by tarruda** [comment]
> I wish I could develop/add on to my favorite editor in js. :)

That will be possible for neovim sooner than you think

**Comment by trusktr** [comment]
That will be a drug with zero negative side effects. :syringe:

**Comment by justinmk** [comment]
Part of this is closed by https://github.com/neovim/neovim/pull/2288

The `OptionSet` thing was added to upstream Vim, so that's also on the way, for better or worse...
